### ISSUE TITLE
Why NOT?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-elm#2]
A few reasons why we might _NOT_ want to use Elm: (_feel free to `add` to this list_!)

+ additional ***learning curve*** for people who have only just learned JavaScript
+ **functional/immutable** paradigm is ***not everyone***'s cup-of-tea
+ it's virtually impossible to find a person with Elm experience on their "CV" (_which means our clients will have to train up anyone they hire to maintain/extend any code we write for them..._)
+ encourages more **front-end-heavy** apps as opposed to progressive web apps
+ may be "_**difficult** to **sell**_" to organisations who have not _heard_ of it. (_most!_)

> _**Note**: some of these `"reasons"` are not `specific` to `elm` but are applicable to `all` compile-to-JS languages e.g. [`clojurescript`](https://github.com/clojure/clojurescript) because by deciding to use them you are forcing `all` the people on your team and client(s) team(s) to learn a whole other language just so they can update a line of code in the codebase...

**Comment by iteles** [comment]
All of the ones on my mind are in your comment above. The really big ones for me is how hard it is to find people who code in elm, which will mean our QA process will suffer of course, but in particular means we leave our clients with something that's *difficult for them to extend* - this goes against our core philosophies. We can provide beautifully composed, commented and documented code, but if they're going to find it either near-impossible or unnecessarily expensive to extend/use further, are we really doing our job to the standard we want it to be?
The other big one for me is the learning curve. Even if it's not *hard*, it forces people into learning *one more thing* which they will o my use in certain cases (if/where appropriate).

**Comment by nelsonic** [comment]
@iteles yeah, I _hear_ what you are saying and I listened _intently_ to what @sofiapoh & @nogainbar said on Friday @ Lunch. Learning JS is _difficult enough_. Then after _barely_ knowing JS we are expected to learn Functional Programming and 20 _other_ things... It's _exhausting_! 😫 
That's _exactly_ why we need to have this list.
And, we need to _consult_ people: https://twitter.com/nelsonic/status/795164336834617344
I would _never_ "_force_" anyone to learn an _entirely_ new programming language just for the sake of it. At this point we are **100% at the _investigation_ phase**. 
If we conclude that elm is a "_too much_" at least we know we gave it a chance. 😉

**Comment by jedwards1211** [comment]
@nelsonic also now that it's possible to run the same JS on both client and server and there are tons and tons and tons of great JS libs in npm, not being able to take advantage of that (though I don't know for sure if it's impossible using Elm) is a disadvantage.

**Comment by nelsonic** [comment]
@jedwards1211 ***`Yes`***, the `promise` of running JS code on both Servers and Client was the _original_ reason we moved to Node.js 🌈 
***`And`*** and until node `V4` we were _able_ to write `universal` code without _needing_ a "_transpiler_" ... 😞 

***`Yes`***, there are _many_ good JS libraries on NPM 📈 
***`And`*** we have _a few_ packages in the "top 1%" of usage ... ***`And`*** we still ❤️  the platform
***`However`***, we don't want to `let` our current "_good_" system _prevent_ us from _exploring_ a _great_ one. 😉 

***`So`*** ... we are _investigating_ our options ⚖ 
***`And`*** _hopefully_ we can make a _proposal_ (_either way_) back to our _team_ for _consideration_... 👍

**Comment by jedwards1211** [comment]
@nelsonic another important question: does Elm provide some way to use duck typing when desired?
Seeing the vitality of the JS ecosystem after coming from Java, I try not to underestimate the value of dynamic typing anymore.

**Comment by nelsonic** [comment]
@jedwards1211 from my _limited_ knowledge I don't think [**`"duck typing"`**](https://en.wikipedia.org/wiki/Duck_typing) is a "_thing_" in Elm ... this is _probably_ worth opening a ***separate issue*** for and investigating it with the help of the people with _way_ more experience than we _currently_ have. 👍  (_**`Yes`**, it could be a "reason" why someone might not consider using `elm-lang` but is it a "**deal-breaker**"...?_)

**Comment by rjmk** [comment]
In languages like Elm, I believe type classes are the closest technique to duck typing. They are not present in Elm. This is a way in which Elm eschews power for approachability.

Some people have found it to be annoying ([1](http://reasonablypolymorphic.com/blog/elm-is-wrong), [2](https://www.youtube.com/watch?v=O_kWwaghZ9U)). It can sometimes be circumvented in interesting ways (e.g. http://package.elm-lang.org/packages/eeue56/elm-all-dict/2.0.0/EveryDict).

So that's one reason why not. [These slides](http://doppioslash.com/slides/codemesh2016.pdf) on problems with Elm are excellent and contain good examples of how Elm's limited power can be frustrating in certain circumstances. Worth noting, there's no implication that JavaScript would be a better choice despite those frustrations

**Comment by jhg** [comment]
Elm hasn't basic features in core development even it hasn't it in comunity development, because they are "conservative". Then for basic things like have a CSS external and work whit designers in the team you need third part libraries, that don't give you guarantee of maintenance in a short future, or you need write your own libraries like anyone.

It need still more development and to be mature, and the first step I think it's accept more easy libraries in the core for things that all people need. One HTML generated from the compiler that nobody can use because it's ugly and you haven't option for add CSS is a bad feature or a incomplete feature in the compiler.

https://github.com/elm-lang/html/issues/113

**Comment by nelsonic** [comment]
@jhg all _great_ points. ✅  (_welcome to DWYL!_) ☀️ 

We _agree_ that HTML output from Elm Compiler is "_unfriendly_". 
I don't think it's _meant_ to be human-readable, _however_ we consider this more of a "tradeoff" than a "flaw". Making _generated_ HTML look good is a "_nice-to-have_" not a "requirement" the Elm Core people need to be _wasting_ their time on at the expense of Server Side Rendering... 😉 
 
If you "View Source" on any Google Page e.g: view-source:https://www.google.co.uk
the HTML+In-Line-JS+CSS is not _human friendly_ because it's (_highly_) Optimised for lowest possible transfer/load time. 

Not saying that is the case with Elm because there are _many_ Language features e.g: [9 Levels of "Currying"](https://github.com/dwyl/learn-elm/blob/cc100bebf52a1f0408bcf1b9366d783a3c314429/index.html#L79) that I _know_ I'm ***never going to use***, which I would prefer to be stripped out. 

> I think I can "_pipe_" the Elm compiler output through an optimiser like https://github.com/dwyl/learn-closure-compiler/issues/1 to trim down the unnecessary functions and make the output _much_ leaner. 

As for not "_allowing_" CSS we're _embracing_ that "_constraint_" and making it a "_feature_" by using Tachyons! see: https://github.com/dwyl/learn-tachyons
This allows us to declare all Markup and Presentation within the Elm Component without fear that it can be "_mutated_" by someone _accidentally_ updating a style somewhere `else`...!

> If it was up-to-me, I would _never_ write another _line_ of CSS again! 😮 

As for not allowing lots of new features into the "_core_", I'm _delighted_ that they are being _ultra-conservative_ with what gets in. I've seen platforms get _unncessarily bloated_ very quickly because too many (_good_) ideas were included at the expense of _few_ ***fantastic*** ones.

![image](https://cloud.githubusercontent.com/assets/194400/22639883/275a10a6-ec48-11e6-9220-c74c076a38c5.png)

> `Babel` is a (_feature_) "_free-for-all_" and `while` it has more _users_ it also has _way more **bugs**_!
`Babel` had to remove their bug-tracking from GitHub cause it was getting _embarrassing_...

I'm _personally_ a _huge fan_ of what Evan has done with Elm. I don't "_love_" ***everything***, but the good is 95% and "bad" only 5% and it's _so much_ better than the alternative! 
I'm _excited_ every time I read Elm Code instead of JS! 😍 

Thanks for your feedback! All valid points everyone should consider before adopting Elm. 👍

**Comment by jhg** [comment]
I agree about remove everything that is not essential, and take care with new features, but some things are basic, if they don't want a function for change the title, ok, take care with that new feature, but explain how Elm allow take any node from the dom and modified it, like do the virtual-dom or other low level core libraries. But not explain how do oneself and don't add basic features is not good.

And a flag for add a CSS to the default compiler is like if when added that feature didn't take care and the feature is unfinish... I don't care if the output of one compiler is not readable by human, I don't want edit that because is the work of the compiler do it, but if when is running is not good for go to production then the flag for add CSS is good idea. Else, make default the output to unuself HTML was a bad idea. And, even the code generate in the HTML is not enought unreadable.

And the same problem from compiler is in reactor, I can test my Elm app, but with all without CSS... when I see it I think "ok, first contact, maybe it is possible but need read more" and not, the web development is not a CLI app, is a GUI then it needs CSS, I think Elm not need worry about generate CSS for to have in CSS the "how to see" and in Elm only the "how to do" even I'm not sure if is good that Elm have a Html module and the view that generate the HTML or better that the view was in a HTML template and Elm only take it from dom and do changes with virtual-dom but without HTML declarations in Elm code.

**Comment by jhg** [comment]
And sincerely, I was excited with Elm, I start thinking not write more JavaScript and write all web application with Elm, write a CSS and, done, with the HTML that the compiler generate, but when I see is not possible do it, that need compile to JavaScript an write JavaScript for make Elm run... the good first contact was broke.

**Comment by nelsonic** [comment]
If you've read _this_ far ...
# GOTO: https://github.com/dwyl/learn-elm/issues/129
For the _reasoning_ why we think Elm is the _best_ choice for front-end web applications in 2019!
